util/dvsim/utils.py

```diff
@@ -86,84 +95,280 @@ def parse_hjson(hjson_file):
-                            # Expand list into a str since list within list is
-                            # not supported.
-                            found = " ".join(subst_found)
-
-                        elif type(found) is str:
-                            found = subst_wildcards(found, mdict,
-                                                    ignored_wildcards,
-                                                    ignore_error)
-
-                        elif type(found) is bool:
-                            found = int(found)
-                        var = var.replace("{" + item + "}", str(found))
-                        no_substitutions_found = False
-                    else:
-                        # Check if the wildcard exists as an environment variable
-                        env_var = os.environ.get(item)
-                        if env_var is not None:
-                            var = var.replace("{" + item + "}", str(env_var))
-                            no_substitutions_found = False
-                        elif not ignore_error:
-                            log.error(
-                                "Substitution for the wildcard \"%s\" not found",
-                                item)
-                            sys.exit(1)
-
-            # If items were found for substitution, but if they were a part
-            # of ignored_wildcards list or if substitutions for them were not
-            # found, then, we return. If all substitutions were made, then check
-            # for second level of indirection:
-            #
-            # For example: lets say we supply the dict:
-            # {var: '{{foo}_xyz{bar}}', foo: p, bar: q, p_xyz_q: baz}
-            #
-            # Then after the substitutions above: {var: '{p_xyz_q}', ...}
-            # We need to now substitute {p_xyz_q}, so that the final value of
-            # var is 'baz'.
-            if not (ignored_wildcards_found or no_substitutions_found):
-                var = subst_wildcards(var, mdict, ignored_wildcards,
-                                      ignore_error)
-    return var
+    wildcard_re = re.compile(r"{([A-Za-z0-9\_]+)}")
+
+    # Work from left to right, expanding each wildcard we find. idx is where we
+    # should start searching (so that we don't keep finding a wildcard that
+    # we've decided to ignore).
+    idx = 0
+
+    any_err = False
+
+    while True:
+        right_str = var[idx:]
+        match = wildcard_re.search(right_str)
+
+        # If no match, we're done.
+        if match is None:
+            return (var, any_err)
+
+        name = match.group(1)
+
+        # If the name should be ignored, skip over it.
+        if name in ignored:
+            idx += match.end()
+            continue
+
+        # If the name has been seen already, we've spotted circular recursion.
+        # If we're ignoring errors, skip over it. Otherwise, error.
+        if name in seen:
+            if ignore_error:
```

Comment:
We are already facing this issue at Google (multiple tools shipping with shared libs that don't play well with each other). IMO this is quite useful. It obviates the need to put any more logic in the Python script. 

This type of construct will be used ONLY for env vars that need to be modified just for running the EDA tool. If the user misuses it, it is their problem. If `LD_LIBRARY_PATH` is not defined, then VCS will not work at all, so you definitely need it to be set before invoking `dvsim`. 

All keys in `exports` are exported to the sub process where the tool is invoked, so it already does not affect the current terminal in any way.

Reply:
Ah, I understand. This still feels awkward, though, because the resulting code looks like
```
exports: [
    ...
    LD_LIBRARY_PATH: "/usr/lib:{LD_LIBRARY_PATH}"
    ...
]
```
and the question of whether this is syntactically well defined would depend on the contents of the process's environment. I just don't think that's the right thing to do!

Thinking about this further, though, the `exports` list is parsed by hjson as a list, not a dictionary (I only just noticed the square brackets above). It only gets turned into a dictionary in `FlowCfg._process_exports`, which happens after substitution. So maybe this doesn't matter at all? I *think* that the existing code and this proposed patch both treat the `LD_LIBRARY_PATH` example in the same way, and act as if they're expanding a string that just happens to look something like `"foo: blah blah {foo}"`.

Reply:
I dont recall why I chose to make `exports` a list of strings. I will push a PR soon to fix that (change it to a list of dicts instead). That should help eliminate `process_exports()`.